01-ruby-fundamentals/hashes-classes-and-you.md

```diff
@@ -1,107 +1,106 @@
-    @first_name  = first_name
-    @last_name   = last_name
-    @street_one  = street_one
-    @street_two  = street_two
-    @city        = city
-    @state       = state
-    @country     = country
-    @postal_code = postal_code
-  end
-end
-
-Address.new("Jeremy","Flores","123 Fake St.","Apt Yes","Seattle","WA","USA","98115")
-```
+We've spent a lot of time talking about classes in recent days. Let's compare them to hashes.
 
-__Question: What happens if you don't have a `street_two` attribute?__
+Classes  | Hashes
+---      | ---
+Use _instance variables_ to store data | Use _key/value_ pairs to store data
+Stored data can be of any type, including other objects and hashes | Stored data can be of any type, including other objects and hashes
+Have methods | Do not have methods
+Make copies with `.new` | Make copies by duplicating all keys / values
+Require a pre-written class definition | Can be created on the fly
 
-Let's try this again. This time, let's initialize using a `Hash`:
+Classes take a little more up-front work. You need to know roughly what your data looks like in advance, and build the class to match. The advantage gained by this effort is that you get methods, including a constructor to quickly make many instances.
 
-```ruby
-class Address
-  def initialize(address_hash)
-    @first_name  = address_hash[:first_name]
-    @last_name   = address_hash[:last_name]
-    @street_one  = address_hash[:street_one]
-    @street_two  = address_hash[:street_two]
-    @city        = address_hash[:city]
-    @state       = address_hash[:state]
-    @country     = address_hash[:country]
-    @postal_code = address_hash[:postal_code]
-  end
-end
-
-Address.new(first_name: "Jeremy", state: "WA")
-```
+Hashes are much more dynamic and flexible. They're simpler and lighter-weight than full classes, making them great when you just want to quickly store some data without a bunch of up-front work. They're also great if you don't know in advance what data you'll be storing, like if you're reading data from disk or from the network.
 
-Using a hash to provide parameters to an `initialize` method lets us omit parameters that may not exist for some instances. It also adds clarity to `class` instantiation by explicitely stating keys and values.
+Choosing how your data will be stored is one of the first steps in writing a complex program. Here are some guidelines:
 
-```ruby
-# not using a hash to initialize
-Address.new("Jeremy","","","","WA","","")
+If your data... | Then you should use a...
+---             | ---
+Comes from an external source, like a file or the internet | Hash
+Involves giving names to pieces of data, without any specific behavior | Hash
+Has complex behaviors you need to model | Class
+Involves many pieces of data following a repeated pattern | Class
 
-# now with a fancy hash!
-Address.new(first_name: "Jeremy", state: "WA")
-```
+<!-- DPR: don't really want to emphasize how confusing things are
```

Comment:
Highlighting that is probably a good idea, but I'm not sure it fits into this resource with the way it's been restructured. It might be a good thing to add to the methods resource.